### PR TITLE
reconfig on deliver

### PR DESCRIPTION
**Pull request by HagarMeir**
Signed-off-by: Hagar Meir <hagar.meir@ibm.com>

### CODE REVIEW

File: pkg/consensus/consensus.go [hunk 1]
```diff
@@ -63,10 +67,15 @@ func (c *Consensus) Complain(viewNum uint64, stopView bool) {
 	c.viewChanger.StartViewChange(viewNum, stopView)
 }
 
-func (c *Consensus) Deliver(proposal types.Proposal, signatures []types.Signature) {
+func (c *Consensus) Deliver(proposal types.Proposal, signatures []types.Signature) bool {
```
Comment (yacovm) [inline pkg/consensus/consensus.go]: Can't we return the reconfig object here? The name of the method is `deliver` so it's not clear from its name, why it returns a boolean and what the boolean means.

File: test/reconfig_test.go [hunk 2]
```diff
@@ -54,17 +54,132 @@ func TestBasicReconfig(t *testing.T) {
...
 	for i := 0; i < numberOfNodes; i++ {
 		d := <-nodes[i].Delivered
 		data = append(data, d)
 	}
 	for i := 0; i < numberOfNodes-1; i++ {
 		assert.Equal(t, data[i], data[i+1])
 	}
+}
+
+func TestBasicAddNode(t *testing.T) {
```
Comment (yacovm) [inline test/reconfig_test.go]: looks good

File: test/test_app.go [hunk 3]
```diff
@@ -73,14 +73,26 @@ func (a *App) Submit(req Request) {
 // Sync synchronizes and returns the latest decision
 func (a *App) Sync() types.SyncResponse {
 	records := a.Node.cb.readAll(*a.latestMD)
+	reconfigSync := types.ReconfigSync{InReplicatedDecisions: false}
```
Comment (yacovm) [inline test/test_app.go:76]: false is the default value, we can just initialize it with `var reconfigSync types.ReconfigSync`

File: pkg/consensus/consensus.go [hunk 4]
```diff
@@ -166,8 +167,14 @@ func (c *Consensus) reconfig(reconfig types.Reconfig) {
 	c.collector.Stop()
 
 	c.Config = reconfig.CurrentConfig
-	if err := c.ValidateConfiguration(); err != nil {
-		c.Logger.Panicf("Configuration is invalid, error: %v", err)
+	if err := c.ValidateConfiguration(reconfig.CurrentNodes); err != nil {
+		if strings.Contains(err.Error(), "Nodes does not contain the SelfID") {
+			c.close()
+			c.Logger.Infof("Closing consensus")
```
Comment (yacovm) [inline pkg/consensus/consensus.go]: let's explain explicitly why we're closing.

File: pkg/consensus/consensus.go [hunk 5]
```diff
@@ -256,26 +263,25 @@ func (c *Consensus) proposalMaker() *algorithm.ProposalMaker {
...
 	if err := c.Config.Validate(); err != nil {
 		return errors.Wrap(err, "bad configuration")
 	}
 
-	nodes := c.Comm.Nodes()
 	nodeSet := make(map[uint64]bool)
 	for _, val := range nodes {
 		if val == 0 {
-			return errors.Errorf("Comm.Nodes() contains node id 0 which is not permitted, nodes: %v", nodes)
+			return errors.Errorf("Nodes contains node id 0 which is not permitted, nodes: %v", nodes)
```
Comment (yacovm) [inline pkg/consensus/consensus.go]: errors start with lowercase

File: test/test_app.go (hunk 3 above)
Comment (HagarMeir) [reply test/test_app.go:76]: It is more clear to me this way, if you don't mind

**Review by yacovm — APPROVED** [review]

**Pull request merged**